syntax = "proto3";

package Acoustid.Server.PB;

message GetAttributeRequest {
  string index_name = 1;
  string name = 2;
}

message GetAttributeResponse {
  string value = 2;
}

message InsertOrUpdateDocumentOp {
  uint32 doc_id = 1;
  repeated uint32 terms = 2;
};

message DeleteDocumentOp {
  uint32 doc_id = 1;
};

message SetAttributeOp {
  string name = 1;
  string value = 2;
};

message Operation {
    oneof op {
        InsertOrUpdateDocumentOp insert_or_update_document = 1;
        DeleteDocumentOp delete_document = 2;
        SetAttributeOp set_attribute = 3;
    }
};

message UpdateRequest {
  string index_name = 1;
  repeated Operation ops = 2;
};

message UpdateResponse {
};

message SearchResult {
    uint32 doc_id = 1;
    float score = 2;
};

message SearchRequest {
    string index_name = 1;
    repeated uint32 terms = 2;
    int32 max_results = 3;
};

message SearchResponse {
    repeated SearchResult results = 1;
};

message GetIndexRequest {
    string index_name = 1;
};

message GetIndexResponse {
};

message CreateIndexRequest {
    string index_name = 1;
};

message CreateIndexResponse {
};

message DeleteIndexRequest {
    string index_name = 1;
};

message DeleteIndexResponse {
};

service Index {
    rpc GetIndex(GetIndexRequest) returns (GetIndexResponse);
    rpc CreateIndex(CreateIndexRequest) returns (CreateIndexResponse);
    rpc DeleteIndex(DeleteIndexRequest) returns (DeleteIndexResponse);

    rpc GetAttribute(GetAttributeRequest) returns (GetAttributeResponse);

    rpc Update(UpdateRequest) returns (UpdateResponse);
    rpc Search(SearchRequest) returns (SearchResponse);
};
